File: app/src/main/java/ru/MeatGames/roguelike/tomb/model/HeroClass.java
package ru.MeatGames.roguelike.tomb.model;

import android.graphics.Bitmap;

import java.util.LinkedList;

import ru.MeatGames.roguelike.tomb.Global;

public class HeroClass {

    // System
    public int x;
    public int y;
    public int mx;
    public int my;
    public Item[] equipmentList;
    public boolean side = true; //left

    // Stats
    public int regen;
    public int cregen;
    public int init = 10;
    public Bitmap[] img = new Bitmap[4];
    public LinkedList<Item> mInventory;

    public void newHero() {
        mInventory = null;
        mInventory = new LinkedList<>();
        equipmentList = new Item[3];
        Global.INSTANCE.getStats()[0].setValue(10);
        Global.INSTANCE.getStats()[1].setValue(10);
        Global.INSTANCE.getStats()[2].setValue(10);
        Global.INSTANCE.getStats()[3].setValue(10);
        Global.INSTANCE.getStats()[4].setValue(10);
        Global.INSTANCE.getStats()[5].setValue(18);
        Global.INSTANCE.getStats()[6].setValue(18);
        Global.INSTANCE.getStats()[7].setValue(10);
        Global.INSTANCE.getStats()[8].setValue(10);
        Global.INSTANCE.getStats()[9].setValue(10);
        Global.INSTANCE.getStats()[10].setValue(10);
        Global.INSTANCE.getStats()[11].setValue(2);
        Global.INSTANCE.getStats()[12].setValue(1);
        Global.INSTANCE.getStats()[13].setValue(3);
        Global.INSTANCE.getStats()[14].setValue(10);
        Global.INSTANCE.getStats()[16].setValue(10);
        Global.INSTANCE.getStats()[18].setValue(1000);
        Global.INSTANCE.getStats()[19].setValue(10);
        Global.INSTANCE.getStats()[20].setValue(0);
        Global.INSTANCE.getStats()[21].setValue(32);
        Global.INSTANCE.getStats()[22].setValue(1);
        Global.INSTANCE.getStats()[25].setValue(10);
        Global.INSTANCE.getStats()[27].setValue(1000);
        Global.INSTANCE.getStats()[28].setValue(10);
        Global.INSTANCE.getStats()[29].setValue(1);
        Global.INSTANCE.getStats()[31].setValue(1);
        cregen = regen = 22;
        addItem(Global.INSTANCE.getGame().createItem(1));
        addItem(Global.INSTANCE.getGame().createItem(4));
        addItem(Global.INSTANCE.getGame().createItem(7));
        addItem(Global.INSTANCE.getGame().createItem(10));
        addItem(Global.INSTANCE.getGame().createItem(10));
        addItem(Global.INSTANCE.getGame().createItem(10));
        for (int i = 0; i < 30; i++) {
            addItem(Global.INSTANCE.getGame().createItem(i % 11));
        }
        preequipItem(mInventory.get(0));
        preequipItem(mInventory.get(1));
        preequipItem(mInventory.get(2));
    }

    public int getStat(int id) {
        return Global.INSTANCE.getStats()[id].getValue();
    }

    public void modifyStat(int id, int value, int m) {
        Global.INSTANCE.getStats()[id].setValue(Global.INSTANCE.getStats()[id].getValue() + m * value);
        if (Global.INSTANCE.getStats()[id].getMaximum() && Global.INSTANCE.getStats()[id].getValue() > Global.INSTANCE.getStats()[id + 1].getValue())
            Global.INSTANCE.getStats()[id].setValue(Global.INSTANCE.getStats()[id + 1].getValue());
        if (id == 20) isLevelUp();
    }

    public void isLevelUp() {
        if (getStat(20) >= getStat(21)) {
            modifyStat(20, getStat(21), -1);
            modifyStat(21, getStat(21), 1);
            modifyStat(31, 1, 1);
            Global.INSTANCE.getMapview().addLine("Уровень повышен!");
            int u = Global.INSTANCE.getGame().rnd.nextInt(3) + 2;
            modifyStat(6, u, 1);
            modifyStat(5, u, 1);
            Global.INSTANCE.getMapview().addLine("Здоровье увеличено");
            if (getStat(31) % 4 == 0) {
                modifyStat(12, 1, 1);
                Global.INSTANCE.getMapview().addLine("Минимальный урон увеличен");
            }
            if (getStat(31) % 5 == 0) {
                regen--;
                if (regen < cregen) cregen = regen;
                Global.INSTANCE.getMapview().addLine("Скорость регенерации увеличена");
            }
            if (getStat(31) % 2 == 0) {
                modifyStat(13, 1, 1);
                Global.INSTANCE.getMapview().addLine("Максиммальный урон увеличен");
            }
            if (getStat(31) % 3 == 0) {
                modifyStat(19, 1, 1);
                Global.INSTANCE.getMapview().addLine("Защита увеличена");
            }
        }
    }

    public void addItem(Item item) {
        mInventory.add(item);
    }

    public boolean isEquiped(Item item) {
        return equipmentList[item.type - 1] == item;
    }

    public void dropItem(Item item) {
        if (!item.isConsumable() && isEquiped(item))
            takeOffItem(item);
        mInventory.remove(item);
        Global.INSTANCE.getMap()[mx][my].addItem(item);
        Global.INSTANCE.getMapview().addLine(item.n + " выброшен" + item.n1);
        Global.INSTANCE.getGame().move(0, 0);
    }

    public void deleteItem(Item item) {
        mInventory.remove(item);
    }

    public void preequipItem(Item item) {
        switch (item.type) {
            case 1:
                equipmentList[0] = item;
                modifyStat(11, item.val1, 1);
                modifyStat(12, item.val2, 1);
                modifyStat(13, item.val3, 1);
                break;
            case 2:
                equipmentList[1] = item;
                modifyStat(19, item.val1, 1);
                modifyStat(22, item.val2, 1);
                break;
            case 3:
                equipmentList[2] = item;
                modifyStat(19, item.val1, 1);
                modifyStat(22, item.val2, 1);
                break;
        }
    }

    public void equipItem(Item item) {
        switch (item.type) {
            case 1:
                equipmentList[0] = item;
                if (item.property && equipmentList[1] != null) {
                    takeOffItem(1);
                }
                modifyStat(11, item.val1, 1);
                modifyStat(12, item.val2, 1);
                modifyStat(13, item.val3, 1);
                break;
            case 2:
                if (equipmentList[0] != null && equipmentList[0].property) {
                    takeOffItem(0);
                }
                equipmentList[1] = item;
                modifyStat(19, item.val1, 1);
                modifyStat(22, item.val2, 1);
                break;
            case 3:
                equipmentList[2] = item;
                modifyStat(19, item.val1, 1);
                modifyStat(22, item.val2, 1);
                break;
        }
        Global.INSTANCE.getMapview().addLine(item.n + " надет" + item.n1);
        Global.INSTANCE.getGame().move(0, 0);
    }

    public void takeOffItem(Item item) {
        switch (item.type) {
            case 1:
                modifyStat(11, item.val1, -1);
                modifyStat(12, item.val2, -1);
                modifyStat(13, item.val3, -1);
                break;
            case 2:
                modifyStat(19, item.val1, -1);
                modifyStat(22, item.val2, -1);
                break;
            case 3:
                modifyStat(19, item.val1, -1);
                modifyStat(22, item.val2, -1);
                break;
        }
        Global.INSTANCE.getHero().equipmentList[item.type - 1] = null;
        Global.INSTANCE.getMapview().addLine(item.n + " снят" + item.n1);
        Global.INSTANCE.getGame().move(0, 0);
    }

    public void takeOffItem(int i) {
        takeOffItem(equipmentList[i]);
    }
}

File: app/src/main/java/ru/MeatGames/roguelike/tomb/model/HeroClass.kt
package ru.MeatGames.roguelike.tomb.model

import android.graphics.Bitmap
import ru.MeatGames.roguelike.tomb.Global
import java.util.*

class HeroClass {

    // System vars
    var x: Float = 0.toFloat()
    var y: Float = 0.toFloat()
    var mx: Int = 0
    var my: Int = 0
    lateinit var equipmentList: Array<Item?>
    var mIsFacingLeft: Boolean = true
    var mIsResting = false
        private set

    // Stats vars
    var regen: Int = 0
    var cregen: Int = 0
    var init = 10
    var img = arrayOfNulls<Bitmap>(4)
    var mInventory: LinkedList<Item>? = null

    fun newHero() {
        mInventory = null
        mInventory = LinkedList<Item>()
        equipmentList = arrayOfNulls<Item>(3)
        Global.stats[0].value = 10
        Global.stats[1].value = 10
        Global.stats[2].value = 10
        Global.stats[3].value = 10
        Global.stats[4].value = 10
        Global.stats[5].value = 18
        Global.stats[6].value = 18
        Global.stats[7].value = 10
        Global.stats[8].value = 10
        Global.stats[9].value = 10
        Global.stats[10].value = 10
        Global.stats[11].value = 2
        Global.stats[12].value = 1
        Global.stats[13].value = 3
        Global.stats[14].value = 10
        Global.stats[16].value = 10
        Global.stats[18].value = 1000
        Global.stats[19].value = 10
        Global.stats[20].value = 0
        Global.stats[21].value = 32
        Global.stats[22].value = 1
        Global.stats[25].value = 10
        Global.stats[27].value = 1000
        Global.stats[28].value = 10
        Global.stats[29].value = 1
        Global.stats[31].value = 1
        regen = 16
        cregen = regen
        addItem(Global.game.createItem(1))
        addItem(Global.game.createItem(4))
        addItem(Global.game.createItem(7))
        addItem(Global.game.createItem(10))
        addItem(Global.game.createItem(10))
        addItem(Global.game.createItem(10))
        for (i in 0..29) {
            addItem(Global.game.createItem(i % 11))
        }
        preequipItem(mInventory!![0])
        preequipItem(mInventory!![1])
        preequipItem(mInventory!![2])
    }

    fun getStat(id: Int) =
            Global.stats[id].value

    fun modifyStat(id: Int, value: Int, m: Int) {
        Global.stats[id].value = Global.stats[id].value + m * value
        if (Global.stats[id].mIsMaximum && Global.stats[id].value > Global.stats[id + 1].value)
            Global.stats[id].value = Global.stats[id + 1].value
        if (id == 20) isLevelUp()
    }

    fun isLevelUp() {
        if (getStat(20) >= getStat(21)) {
            modifyStat(20, getStat(21), -1)
            modifyStat(21, getStat(21), 1)
            modifyStat(31, 1, 1)
            Global.mapview.addLine("Уровень повышен!")
            val u = Global.game.mRandom.nextInt(3) + 2
            modifyStat(6, u, 1)
            modifyStat(5, u, 1)
            Global.mapview.addLine("Здоровье увеличено")
            if (getStat(31) % 4 == 0) {
                modifyStat(12, 1, 1)
                Global.mapview.addLine("Минимальный урон увеличен")
            }
            if (getStat(31) % 5 == 0) {
                regen--
                if (regen < cregen) cregen = regen
                Global.mapview.addLine("Скорость регенерации увеличена")
            }
            if (getStat(31) % 2 == 0) {
                modifyStat(13, 1, 1)
                Global.mapview.addLine("Максиммальный урон увеличен")
            }
            if (getStat(31) % 3 == 0) {
                modifyStat(19, 1, 1)
                Global.mapview.addLine("Защита увеличена")
            }
        }
    }

    fun addItem(item: Item) =
            mInventory!!.add(item)

    fun isEquipped(item: Item) =
            equipmentList[item.mType - 1] == item

    fun dropItem(item: Item) {
        if (!item.isConsumable && isEquipped(item)) {
            takeOffItem(item)
        }
        mInventory!!.remove(item)
        Global.map!![mx][my].addItem(item)
        Global.mapview.addLine(item.mTitle + " выброшен" + item.mTitleEnding)
        Global.game.skipTurn()
    }

    fun deleteItem(item: Item) =
            mInventory!!.remove(item)

    fun preequipItem(item: Item) {
        when (item.mType) {
            1 -> {
                equipmentList[0] = item
                modifyStat(11, item.mValue1, 1)
                modifyStat(12, item.mValue2, 1)
                modifyStat(13, item.mValue3, 1)
            }
            2 -> {
                equipmentList[1] = item
                modifyStat(19, item.mValue1, 1)
                modifyStat(22, item.mValue2, 1)
            }
            3 -> {
                equipmentList[2] = item
                modifyStat(19, item.mValue1, 1)
                modifyStat(22, item.mValue2, 1)
            }
        }
    }

    fun equipItem(item: Item) {
        when (item.mType) {
            1 -> {
                equipmentList[0] = item
                if (item.mProperty && equipmentList[1] != null) {
                    takeOffItem(1)
                }
                modifyStat(11, item.mValue1, 1)
                modifyStat(12, item.mValue2, 1)
                modifyStat(13, item.mValue3, 1)
            }
            2 -> {
                if (equipmentList[0] != null && equipmentList[0]!!.mProperty) {
                    takeOffItem(0)
                }
                equipmentList[1] = item
                modifyStat(19, item.mValue1, 1)
                modifyStat(22, item.mValue2, 1)
            }
            3 -> {
                equipmentList[2] = item
                modifyStat(19, item.mValue1, 1)
                modifyStat(22, item.mValue2, 1)
            }
        }
        Global.mapview.addLine(item.mTitle + " надет" + item.mTitleEnding)
        Global.game.skipTurn()
    }

    fun takeOffItem(item: Item) {
        when (item.mType) {
            1 -> {
                modifyStat(11, item.mValue1, -1)
                modifyStat(12, item.mValue2, -1)
                modifyStat(13, item.mValue3, -1)
            }
            2 -> {
                modifyStat(19, item.mValue1, -1)
                modifyStat(22, item.mValue2, -1)
            }
            3 -> {
                modifyStat(19, item.mValue1, -1)
                modifyStat(22, item.mValue2, -1)
            }
        }
        Global.hero!!.equipmentList[item.mType - 1] = null
        Global.mapview.addLine(item.mTitle + " снят" + item.mTitleEnding)
        Global.game.skipTurn()
    }

    fun takeOffItem(i: Int) =
            takeOffItem(equipmentList[i]!!)

    @JvmOverloads
    fun startResting(loudBroadcast: Boolean = true) {
        mIsResting = true
        if (loudBroadcast) {
            Global.mapview.addLine("Отдых начат")
        }
    }

    @JvmOverloads
    fun interruptResting(loudBroadcast: Boolean = true) {
        if (mIsResting) {
            mIsResting = false
            if (loudBroadcast) {
                Global.mapview.addLine("Отдых прерван!")
            }
        }
    }

    @JvmOverloads
    fun finishResting(loudBroadcast: Boolean = true) {
        if (mIsResting) {
            mIsResting = false
            if (loudBroadcast) {
                Global.mapview.addLine("Отдых завершен!")
            }
        }
    }

    fun isFullyHealed() =
            getStat(5) == getStat(6)

    // used for proper handling all ongoing hero actions
    // for example when changing screens
    fun interruptAllActions() {
        interruptResting(false)
    }

}